src/sdam/monitor.ts

```diff
@@ -484,61 +478,57 @@ export interface MonitorIntervalOptions {
 export class MonitorInterval {
   fn: (callback: Callback) => void;
   timerId: NodeJS.Timeout | undefined;
-  lastCallTime: number;
-  isExpeditedCheckScheduled = false;
+  lastExecutionEnded: number;
+  isExpeditedCallToFnScheduled = false;
   stopped = false;
+  isExecutionInProgress = false;
+  hasExecutedOnce = false;
 
   heartbeatFrequencyMS: number;
   minHeartbeatFrequencyMS: number;
-  clock: () => number;
 
   constructor(fn: (callback: Callback) => void, options: Partial<MonitorIntervalOptions> = {}) {
     this.fn = fn;
-    this.lastCallTime = 0;
+    this.lastExecutionEnded = 0;
 
     this.heartbeatFrequencyMS = options.heartbeatFrequencyMS ?? 1000;
     this.minHeartbeatFrequencyMS = options.minHeartbeatFrequencyMS ?? 500;
-    this.clock = typeof options.clock === 'function' ? options.clock : now;
 
     if (options.immediate) {
       this._executeAndReschedule();
     } else {
-      this.lastCallTime = this.clock();
+      this.lastExecutionEnded = now();
       this._reschedule(undefined);
     }
   }
 
   wake() {
-    const currentTime = this.clock();
-    const nextScheduledCallTime = this.lastCallTime + this.heartbeatFrequencyMS;
-    const timeUntilNextCall = nextScheduledCallTime - currentTime;
-
-    // For the streaming protocol: there is nothing obviously stopping this
-    // interval from being woken up again while we are waiting "infinitely"
-    // for `fn` to be called again`. Since the function effectively
-    // never completes, the `timeUntilNextCall` will continue to grow
-    // negatively unbounded, so it will never trigger a reschedule here.
-
-    // This is possible in virtualized environments like AWS Lambda where our
-    // clock is unreliable. In these cases the timer is "running" but never
-    // actually completes, so we want to execute immediately and then attempt
-    // to reschedule.
-    if (timeUntilNextCall < 0) {
+    const currentTime = now();
+    const timeSinceLastCall = currentTime - this.lastExecutionEnded;
+
+    if (!this.hasExecutedOnce) {
```

Comment:
1) why do we specifically need this `hasExecutedOnce` property? won't we already hit the executeAndReschedule logic at the bottom of this function if this is the very first time it's called? I don't see any early returns that would prevent it.

2) Do we have a test case for unreliable clocks where we get a timeSinceLastCall is negative? We'd end up rescheduling for an interval greater than `minHeartbeatFrequencyMS`, which isn't what we want. Or are we confident we don't want to do anything about that case?

Reply:
@dariakp 

1. If the function is woken before the function executes but is woken less than minHeartbeatFrequencyMS after it was initialized, we want to execute immediately.  If we remove this check, we'd fall through to `timeSinceLastCall < minHeartbeatFrequencyMS` and reschedule the call instead of executing it immediately.
2. Since the logic has been flipped, I don't think this is relevant anymore.  The `unreliable clocks` doesn't refer to `now`, it refers to `setTimeout` (and node timers) being unreliable.  Previously, it was possible that the timer didn't fire and consequently, `timeUntilNextCall` would be negative.  We now keep track of the time since last call, which will never be negative.  Every time this function is woken, `currentTime` will be calculated and will always be >= the last call time.